convert2rhel/applock.py

```diff
@@ -0,0 +1,171 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import atexit
+import errno
+import io
+import logging
+import os
+import time
+
+
+loggerinst = logging.getLogger(__name__)
+
+
+class ApplicationLockedError(Exception):
+    """Raised when this application is locked."""
+
+    def __init__(self, message):
+        super(ApplicationLockedError, self).__init__(message)
+        self.message = message
+
+
+class ApplicationLock(object):
+    """Holds a lock for a particular application.
+
+    The implementation uses a standard Linux PID file, though that
+    fact (and all implementation details) are hidden from the caller.
+    To acquire and release the lock, use trylock() and unlock().
+    """
+
+    def __init__(self, name):
+        # Our application name
+        self._name = name
+        # Directory in which the lockfile will be written
+        self._lock_dir = "/var/run/lock"
+        # Do we think we locked the pid file?
+        self._locked = False
+        # Maximum number of tries to lock
+        self._max_loop_count = 5
+        # Our process ID
+        self._pid = os.getpid()
+        # Path to the file that contains the process id
+        self._pidfile = os.path.join(self._lock_dir, self._name + ".pid")
+
+    def __str__(self):
+        if self._locked:
+            status = "locked"
+        else:
+            status = "unlocked"
+        return "%s PID %d %s" % (self._pidfile, self._pid, status)
+
+    def set_lock_dir(self, new_tmp_dir):
+        """Set the lock directory. Used only for testing.
+
+        :param new_tmp_dir: the directory to be used for the lock file
+        """
+        self._lock_dir = new_tmp_dir
+        self._pidfile = os.path.join(self._lock_dir, self._name + ".pid")
+
+    def _try_creat(self):
+        """Try to create the lock file. If this succeeds, the lock file
+        exists and we created it.
+
+        :returns: True if we created the lock, False if we didn't.
+        """
+        try:
+            file_desc = os.open(self._pidfile, os.O_CREAT | os.O_WRONLY | os.O_EXCL, 0o755)
+        except OSError as exc:
+            if exc.errno == errno.EEXIST:
+                return False
+            raise exc
+        encoded = (str(self._pid) + "\n").encode("ascii")
+        os.write(file_desc, encoded)
+        os.close(file_desc)
+        loggerinst.debug("%s." % self)
+        return True
+
+    def is_locked(self):
+        """Test whether this object is locked."""
+        return self._locked
+
+    @staticmethod
+    def _pid_exists(pid):
+        """Test whether a particular process exists."""
+        try:
+            os.kill(pid, 0)
+        except ProcessLookupError:
+            return False
+        except PermissionError:
+            pass
```

Comment:
Maybe a little comment here explaining with there is a permission error. Just to make code scanning happy, as this is easily doable. 

Also, might not be intuitive to everyone why PermissionError is raised here when checking for PIDs, so that might be useful to other people reading the code at some point.